### PR TITLE
ROC 5240 directions questionnaire feature toggle and dashboard 

### PR DESCRIPTION
**Pull request by tharacka**
### JIRA link

https://tools.hmcts.net/jira/browse/ROC-5240

### Change description
changes to add a FF4J feature toggle
changes to add a local feature toggle
changes to add changes to task builder (response and claimant response)
changes to folder structure to add a new section for directions questionnaire

### Work checklist

- [ ] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [ ] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/features/directions-questionnaire/index.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  return AuthorizationMiddleware.requestHandler(requiredRoles, accessDeniedCallback, unprotectedPaths)
+}
+
+export class DirectionsQuestionnaireFeature {
+  enableFor (app: express.Express) {
+    const allDQs = '/case/*/directions-questionnaire/*'
+    app.all(allDQs, requestHandler())
+    app.all(allDQs, ClaimMiddleware.retrieveByExternalId)
+    app.all(allDQs, CountyCourtJudgmentRequestedGuard.requestHandler)
+    app.all(allDQs,
```
Comment (ellie-harrison) [inline src/main/features/directions-questionnaire/index.ts]: Quick thought, we should probably have a guard here to check that the claim has the new 'directions-questionnaire' feature on it

File: src/main/features/claim/routes/pay.ts [hunk 2]
```diff
@@ -93,11 +93,23 @@ async function successHandler (res, next) {
...
+    let features: string
+    if (await featureTogglesClient.isFeatureToggleEnabled(user, roles, 'cmc_admissions')) {
+      features = 'admissions,'
+    }
+
+    if (await featureTogglesClient.isFeatureToggleEnabled(user, roles, 'cmc_directions_questionnaire')) {
+      features += 'directions_questionnaire'
     }
+
+    // if (await featureTogglesClient.isAdmissionsAllowed(user, roles)) {
```
Comment (sabahirfan) [inline src/main/features/claim/routes/pay.ts]: If this code is not required, please remove these lines.

File: src/main/features/claimant-response/helpers/taskListBuilder.ts [hunk 3]
```diff
@@ -291,4 +294,23 @@ export class TaskListBuilder {
...
+  static buildDirectionsQuestionnaireSection (draft: DraftClaimantResponse,
+                                              claim: Claim,
+                                              directionsQuestionnaireDraft?: DirectionsQuestionnaireDraft): TaskList {
+    if (FeatureToggles.isEnabled('directionsQuestionnaire')) {
+      let path: string
+      path = DirectionsQuestionnairePaths.hearingLocationPage.evaluateUri({ externalId: claim.externalId })
+      return new TaskList(
+        'Tell us more about the claim', [
+          new TaskListItem(
+            'Give us details in case there\'s a hearing',
```
Comment (sabahirfan) [inline src/main/features/claimant-response/helpers/taskListBuilder.ts]: Please use correct apostrophe here

File: src/main/features/directions-questionnaire/routes/hearing-location.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  res.render(Paths.hearingLocationPage.associatedView)
+}
+
+/* tslint:disable:no-default-export */
+export default express.Router()
+  .get(Paths.hearingLocationPage.uri, (req: express.Request, res: express.Response) => {
+    renderView(res)
+  })
+  // .post(
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/routes/hearing-location.ts]: If this is redundant code, please remove.

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+{% from "form.njk" import csrfProtection, submitButton %}
+
+{% set heading = 'Choose a hearing location' %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+     <p>Tell us your preferred location for a hearing, in case the claim goes to one.</p>
+
+               <p>We'll use the location you choose unless the claimant has an exceptional reason to object, for example accessibility needs.</p>
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please use correct apostrophe i.e. `We’ll`

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please remove extra lines

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+
+               <p>You can choose any location - Guildford County Court is nearest to the home address you gave us.</p>
+
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please remove extra lines

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                   <div class="panel panel-border-narrow js-hidden" id="postcode">
+                     <div class="form-group">
+                       <label class="form-label" for="county_court">Postcode</label>
+                       <input type="text" id="county_court" name="county_court" class="form-control" style="width: 120px" />
+                     </div>
+                   </div>
+                 </fieldset>
+               </div>
+
+
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please remove extra lines

File: src/main/features/response/helpers/taskListBuilder.ts [hunk 9]
```diff
@@ -269,6 +272,23 @@ export class TaskListBuilder {
...
   }
 
+  static buildDirectionsQuestionnaireSection (draft: ResponseDraft, claim: Claim, directionsQuestionnaireDraft?: DirectionsQuestionnaireDraft): TaskList {
+    if (FeatureToggles.isEnabled('directionsQuestionnaire')) {
+      let path: string
+      path = DirectionsQuestionnairePaths.hearingLocationPage.evaluateUri({ externalId: claim.externalId })
+      return new TaskList(
+      'Tell us more about the claim', [
+        new TaskListItem(
+          'Give us details in case there\'s a hearing',
```
Comment (sabahirfan) [inline src/main/features/response/helpers/taskListBuilder.ts]: Please use correct apostrophe

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 10]
```diff
@@ -0,0 +1,59 @@
+{% extends "layout.njk" %}
+{% from "form.njk" import csrfProtection, submitButton %}
+
+{% set heading = 'Choose a hearing location' %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <p>Tell us your preferred location for a hearing, in case the claim goes to one.</p>
```
Comment (robertparkinson) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: This entire file is missing translations

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      <p>We’ll use the location you choose unless the claimant has an exceptional reason to object, for example accessibility needs.</p>
+      <p>You can choose any location - Guildford County Court is nearest to the home address you gave us.</p>
+
+      <h1 class="heading-medium">Do you want to choose Guildford County Court?</h1>
+
+      <div class="form-group">
+        <fieldset>
+         <div class="multiple-choice">
```
Comment (robertparkinson) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please use the yesNoRadio

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        <h1 class="heading-medium">Choose another court</h1>
+        <fieldset>
+         <div class="multiple-choice" data-target="name">
+           <input id="radio-court-name" type="radio" name="court-search" value="name">
+           <label for="radio-court-name">Enter court name</label>
+         </div>
+         <div class="panel panel-border-narrow js-hidden" id="name">
+           <div class="form-group">
+             <label class="form-label" for="county_court">Court name</label>
+             <input type="text" id="county_court" name="county_court" class="form-control" style="width: 500px" />
```
Comment (robertparkinson) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: don't add style to inputs. please use classes. there should already be existing class to use

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+           </div>
+         </div>
+         <div class="multiple-choice" data-target="postcode">
+           <input id="radio-court-postcode" type="radio" name="court-search" value="postcode">
+           <label for="radio-court-postcode">Search by postcode</label>
+         </div>
+         <div class="panel panel-border-narrow js-hidden" id="postcode">
+           <div class="form-group">
+             <label class="form-label" for="county_court">Postcode</label>
+             <input type="text" id="county_court" name="county_court" class="form-control" style="width: 120px" />
```
Comment (robertparkinson) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: remove style and replace with class

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+           </div>
+         </div>
+        </fieldset>
+      </div>
+
+      <div class="form-group">
+        <input type="submit" class="button" value="Save and continue">
```
Comment (robertparkinson) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: use submitButton macro

File: src/main/features/response/routes/task-list.ts [hunk 15]
```diff
@@ -9,13 +9,15 @@ import { ResponseDraft } from 'response/draft/responseDraft'
 import { Draft } from '@hmcts/draft-store-client'
 import { MomentFactory } from 'shared/momentFactory'
 import { MediationDraft } from 'mediation/draft/mediationDraft'
+// import { DirectionsQuestionnaireDraft } from 'directions-questionnaire/draft/directionsQuestionnaireDraft'
```
Comment (robertparkinson) [inline src/main/features/response/routes/task-list.ts]: does this need removing?

File: src/main/features/response/routes/task-list.ts [hunk 16]
```diff
@@ -9,13 +9,15 @@ import { ResponseDraft } from 'response/draft/responseDraft'
...
 
 /* tslint:disable:no-default-export */
 export default express.Router()
   .get(Paths.taskListPage.uri, async (req: express.Request, res: express.Response, next: express.NextFunction) => {
     try {
       const draft: Draft<ResponseDraft> = res.locals.responseDraft
       const draftMediation: Draft<MediationDraft> = res.locals.mediationDraft
+      // const draftDirectionsQuestionnaire: Draft<DirectionsQuestionnaireDraft> = res.locals.directionsQuestionnaireDraft
```
Comment (robertparkinson) [inline src/main/features/response/routes/task-list.ts]: why is this commented out?

File: src/main/features/claim/routes/pay.ts [hunk 17]
```diff
@@ -93,11 +93,16 @@ async function successHandler (res, next) {
       logger.error(`missing consent not given role for user, User Id : ${user.id}`)
     }
 
-    if (await featureTogglesClient.isAdmissionsAllowed(user, roles)) {
-      await claimStoreClient.saveClaim(draft, user, 'admissions')
-    } else {
-      await claimStoreClient.saveClaim(draft, user)
+    let features: string
+    if (await featureTogglesClient.isFeatureToggleEnabled(user, roles, 'cmc_admissions')) {
+      features = 'admissions,'
```
Comment (robertparkinson) [inline src/main/features/claim/routes/pay.ts]: as discussed logic here for how the comma is handled needs to be looked at

File: src/main/features/directions-questionnaire/views/hearing-location.njk [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+{% from "form.njk" import csrfProtection, submitButton %}
+
+{% set heading = 'Choose a hearing location' %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <p>Tell us your preferred location for a hearing, in case the claim goes to one.</p>
+      <p>We’ll use the location you choose unless the claimant has an exceptional reason to object, for example accessibility needs.</p>
+      <p>You can choose any location - Guildford County Court is nearest to the home address you gave us.</p>
```
Comment (sabahirfan) [inline src/main/features/directions-questionnaire/views/hearing-location.njk]: Please use translation `t('....')` for above three

File: src/main/features/response/routes/task-list.ts (hunk 16 above)
Comment (sabahirfan) [inline src/main/features/response/routes/task-list.ts]: Is this required?

File: src/main/features/response/routes/task-list.ts (hunk 15 above)
Comment (sabahirfan) [inline src/main/features/response/routes/task-list.ts]: Is this required?

File: src/main/features/directions-questionnaire/views/hearing-location.njk (hunk 14 above)
Comment (tharacka) [reply src/main/features/directions-questionnaire/views/hearing-location.njk]: njk removed

File: src/main/features/directions-questionnaire/views/hearing-location.njk (hunk 18 above)
Comment (tharacka) [reply src/main/features/directions-questionnaire/views/hearing-location.njk]: njk removed as it is not required for this story

File: src/main/features/directions-questionnaire/views/hearing-location.njk (hunk 13 above)
Comment (tharacka) [reply src/main/features/directions-questionnaire/views/hearing-location.njk]: njk removed

File: src/main/features/directions-questionnaire/views/hearing-location.njk (hunk 12 above)
Comment (tharacka) [reply src/main/features/directions-questionnaire/views/hearing-location.njk]: njk removed

**Pull request merged**